File: vendour_marketplace_backend/src/main/java/com/vendor_marketplace/entity/Coupon.java
package com.vendor_marketplace.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.proxy.HibernateProxy;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Builder
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true)
    private String code;
    private double discountInPercentage;
    private LocalDateTime startDate;
    private LocalDateTime endDate;
    private double minimumOrderValue;

    @Builder.Default
    private boolean isActive = true;

    @Builder.Default
    @ManyToMany(mappedBy = "usedCoupons")
    private Set<User> usedByUsers = new HashSet<>();

    @CreationTimestamp
    private LocalDateTime createdAt;

}

File: vendour_marketplace_frontend/src/types/coupon.ts
import type { UserResponse } from "./user";

export interface CouponResponse {
    couponId: number;
    code: string;
    discountInPercentage: number;
    startDate: string;          // LocalDate from Java → ISO string in JSON
    endDate: string;
    minimumOrderValue: number;
    active: boolean;
    usedByUsers: UserResponse[];
}

export interface CouponRequest {
    minimumOrderValue: number;
    startDate: string | null;
    endDate: string | null;
    code: string | null;
    isActive?: boolean;
    discountInPercentage: number;
}


// Admin functionality
export interface CouponSliceState {
    coupons: CouponResponse[];
    loading: boolean;
    coupon: CouponResponse | null;
    error: string | undefined | null;
    isCouponCreated: boolean;
}
